tests/ftqc/test_parametric_mid_measure.py

```diff
@@ -624,154 +623,127 @@ def test_diagonalize_mcm_transform_preserves_reset(self, reset):
-        assert len(new_tape.operations) == 6
-        for op in new_tape.operations[3:]:
-            assert isinstance(op, qml.ops.Conditional)
+        assert len(new_tape.operations) == 5
 
         new_mp = new_tape.operations[2]
         assert not isinstance(new_mp, ParametricMidMeasureMP)
         assert isinstance(new_mp, MidMeasureMP)
 
-        assert new_tape.operations[3].meas_val.measurements == [new_mp]
+        # the conditionals' MeasurementValues are mapped to the new, diagonalized mp
+        processing_fns = []
+        for op in new_tape.operations[3:]:
+            assert isinstance(op, qml.ops.Conditional)
+            assert op.meas_val.measurements == [new_mp]
+            processing_fns.append(op.meas_val.processing_fn)
+
+        # true and false processing fns are preserved on the conditionals
+        assert [fn(0) for fn in processing_fns] == [True, False]
+        assert [fn(1) for fn in processing_fns] == [False, True]
 
-    def test_diagonalizing_measurements_cond(self):
-        """Test that when calling diagonalize_mcms, references to previously
-        diagonalized measurements that are stored in conditions on Conditional
+    def test_diagonalizing_mcm_used_as_cond_and_op(self):
+        """Test that when calling diagonalize_mcms, references to diagonalized
+        measurements that are stored in conditions on subsequent Conditional
         operators are updated to track the measurement on the tape following
         diagonalization, rather than the original object (for conditional
-        with MCM condition and non-MCM op)"""
+        with MCM condition and MCM applied op)"""
 
         with qml.queuing.AnnotatedQueue() as q:
             qml.RX(1.2, 0)
             mp = ParametricMidMeasureMP(0, angle=1.2, plane="YZ")
             mv = MeasurementValue([mp], processing_fn=lambda v: v)
-            qml.cond(mv == 0, partial(qml.RX, 1.2))(wires=2)
+            # using cond_meas and conditionally applying measurements
+            cond_meas(mv == 0, measure_x, measure_y)(2)
 
         original_tape = qml.tape.QuantumScript.from_queue(q)
         old_mp = original_tape.operations[1]
         assert isinstance(old_mp, ParametricMidMeasureMP)
 
+        # ops are: RX, diag_gate, mp, conditional(H), conditional(H), conditional(adjoint(S)), mp
         (new_tape,), _ = diagonalize_mcms(original_tape)
-        assert len(new_tape.operations) == 4
-        for op in new_tape.operations[3:]:
-            assert isinstance(op, qml.ops.Conditional)
+        assert len(new_tape.operations) == 7
 
         new_mp = new_tape.operations[2]
         assert not isinstance(new_mp, ParametricMidMeasureMP)
         assert isinstance(new_mp, MidMeasureMP)
 
-        assert new_tape.operations[3].meas_val.measurements == [new_mp]
+        # the conditionals' MeasurementValues are mapped to the new, diagonalized mp
+        processing_fns = []
+        for op in new_tape.operations[3:6]:
```

Comment:
Any concerns with these indices introducing failures with changes to the rest of the ecosystem?

Reply:
You mean if something else changes, this test might fail for reasons unrelated to how the transform works?

That has happened occasionally in other parts of the test suite, with either indices or even just `len(tape)` checks, but I think those are usually tests that integrate a qnode, so "extra" things can happen more easily. Here we're directly creating the tape and then applying _only_ this transform to it, so unless something drastic happens to `AnnotatedQueue` (which would affect things all over the test suite), I would be surprised if we have an issue here.